experiments/tf_trainer/common/cnn_spec_parser.py

```diff
@@ -0,0 +1,87 @@
+# coding=utf-8
+# Copyright 2018 The Conversation-AI.github.io Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""CNN Specification Parser.
+
+A simple parser for specifications of convolutional layers.
+
+BNF defining the syntax to specify CNNs:
+```
+  layers = layer : layers
+  layer = filters
+  filters = filter, filters
+  filter = (size / stride -> num_filters)
+  size, stride, num_filters = \d+
+```
+
+Note that num_filters is the output embedding size.
+"""
+
+import re
+from typing import List
+
+
+layers_split_regexp = re.compile(r'\s*:\s*')
+filters_split_regexp = re.compile(r'\s*,\s*')
+filter_regexp = re.compile(
+  r'\(\s*(?P<size>\d+)\s*/\s*(?P<stride>\d+)\s*'
+  r'\-\>\s*(?P<num_filters>\d+)\s*\)')
+
+
+class FilterParseError(Exception):
+  pass
+
+
+class Filter(object):
+  """A single CNN filter.
+
+  filter = '(size / stride -> num_filters)'
```

Comment:
Why did we choose this notation 'size / stride -> num_filters'? Is there a standard or some reasoning behind it? Perhaps we can add some more information to the PR description to clarify the syntax.

Reply:
I don't know of a standard. I wanted something that I could interpret somehow... here was my thinking... 

The num filters is the final output dimension size, so I put it on the right of the arrow.
(input-length * size / stride) give you the other dimension for the matrix. 
I wanted to have some kind of expression of the input going to the output. 

I think this will have to change a little as I add max-pool vs avg-pool vs other aggregation functions. Maybe "a{ size / stride -> num filters }", or something. 

I'll add a little explanation for this. Its quite easy to change if we come up with a better notation.